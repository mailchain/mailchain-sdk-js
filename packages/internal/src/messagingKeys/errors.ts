export class MessagingKeyContactError extends Error {
	readonly type = 'messaging_key_contact_call_error';
	readonly docs = 'https://docs.mailchain.com/developer/errors/codes#messaging_key_contact_call_error';
	constructor(cause: Error) {
		super(`Failed calling messaging key contract.`, { cause });
	}
}

export class UnexpectedMailchainError extends Error {
	readonly type = 'unexpected_error';
	readonly docs = 'https://docs.mailchain.com/developer/errors/codes#unexpected_error';
	constructor(message: string, cause: any) {
		super(message, { cause });
	}
}

export class IdentityProviderUnsupportedError extends Error {
	readonly type = 'identity_provider_unsupported';
	readonly docs = 'https://docs.mailchain.com/developer/errors/codes#identity_provider_unsupported';
	constructor() {
		super(`Identity provider is not support. Check list of supported providers.`);
	}
}
export class IdentityNotFoundError extends Error {
	readonly type = 'identity_not_found';
	readonly docs = 'https://docs.mailchain.com/developer/errors/codes#identity_not_found';
	constructor() {
		super(`Identity is not found. Check address exists on supplied identity provider.`);
	}
}

export class AddressInvalidError extends Error {
	readonly type = 'address_invalid';
	readonly docs = 'https://docs.mailchain.com/developer/errors/codes#address_invalid';
	constructor(cause: Error) {
		super(`Address is invalid. Check address format, then try again.`, { cause });
	}
}

export class MessagingKeyNotRegisteredError extends Error {
	readonly type = 'messaging_key_unregistered';
	readonly docs = 'https://docs.mailchain.com/developer/errors/codes#messaging_key_unregistered';
	constructor() {
		super('Messaging key is not registered. Register address, then try again.');
	}
}
