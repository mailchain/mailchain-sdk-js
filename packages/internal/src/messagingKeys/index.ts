export * from './addressNonce';
export * from './errors';
export * from './messagingKeys';
export * from './messagingKeys';
export * from './privateMessagingKeys';
export * from './proof';
