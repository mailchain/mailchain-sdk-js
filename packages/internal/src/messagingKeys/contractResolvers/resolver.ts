import { ProtocolType } from '@mailchain/addressing';
import { ContractCall } from '@mailchain/api';
import { PublicKey } from '@mailchain/crypto';
import { MailchainResult } from '../../';
import { Proof } from '../proof';
import { InvalidContractResponseError, MessagingKeyNotFoundInContractError } from './errors';

export type ContractMessagingKey = {
	messagingKey: PublicKey;
	protocol: ProtocolType;
	proof: Proof;
};
export type ContractMessagingKeyError = MessagingKeyNotFoundInContractError | InvalidContractResponseError;
export type ContractCallResolveResult = MailchainResult<
	ContractMessagingKey,
	MessagingKeyNotFoundInContractError | InvalidContractResponseError
>;

export interface ContractCallMessagingKeyResolver {
	resolve(contract: ContractCall): Promise<ContractCallResolveResult>;
}

export interface ContractCallLatestNonce {
	latestNonce(contract: ContractCall): Promise<number>;
}
