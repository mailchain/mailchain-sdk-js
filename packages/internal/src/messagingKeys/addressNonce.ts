import {
	ETHEREUM,
	MAILCHAIN,
	NEAR,
	ProtocolNotSupportedError,
	ProtocolType,
	TEZOS,
	isBlockchainProtocolEnabled,
} from '@mailchain/addressing';
import axios, { AxiosInstance } from 'axios';
import { MessagingKeysApiFactory, MessagingKeysApiInterface, createAxiosConfiguration } from '@mailchain/api';
import { Configuration } from '../configuration';
import { MailchainResult } from '../mailchainResult';
import { ContractCallLatestNonce } from './contractResolvers/resolver';
import { NearContractCallResolver } from './contractResolvers/near';
import { MailchainKeyRegContractCallResolver } from './contractResolvers/mailchain';
import { MessagingKeyNotFoundInContractError } from './contractResolvers/errors';
import { UnexpectedMailchainError } from './errors';

export type GetMessagingKeyLatestNonceResult = MailchainResult<number, GetMessagingKeyLatestNonceError>;
export type GetMessagingKeyLatestNonceError = ProtocolNotSupportedError | UnexpectedMailchainError;

export class AddressNonce {
	constructor(
		private readonly messagingKeysApi: MessagingKeysApiInterface,
		private readonly resolvers: Map<ProtocolType, ContractCallLatestNonce>,
	) {}

	static create(configuration: Configuration, axiosInstance: AxiosInstance = axios.create()) {
		const mailchainKeyRegistryResolver = MailchainKeyRegContractCallResolver.create(configuration, axiosInstance);
		return new AddressNonce(
			MessagingKeysApiFactory(createAxiosConfiguration(configuration.apiPath)),
			new Map<ProtocolType, ContractCallLatestNonce>([
				[NEAR, NearContractCallResolver.create(configuration, axiosInstance)],
				[ETHEREUM, mailchainKeyRegistryResolver],
				[TEZOS, mailchainKeyRegistryResolver],
			]),
		);
	}

	/**
	 * Get the latest nonce for an address.
	 *
	 * @param address the protocol get the latest nonce for.
	 * @param protocol where to find the address.
	 * @returns The latest nonce for the given address.
	 */
	async getMessagingKeyLatestNonce(
		address: string,
		protocol: ProtocolType,
	): Promise<GetMessagingKeyLatestNonceResult> {
		if (protocol === MAILCHAIN) {
			return { data: 1 }; // currently mailchain accounts do not support incrementing nonces
		}

		const resolver = this.resolvers.get(protocol);
		if (!resolver) {
			return { error: new ProtocolNotSupportedError(protocol) };
		}

		if (!isBlockchainProtocolEnabled(protocol)) {
			return { error: new ProtocolNotSupportedError(protocol) };
		}

		const nonceContractResponse = await this.messagingKeysApi.getProtocolAddressNonce(address, protocol);

		try {
			const nonce = await resolver.latestNonce(nonceContractResponse.data.contractCall);
			return {
				data: nonce,
			};
		} catch (error) {
			if (error instanceof MessagingKeyNotFoundInContractError) {
				return {
					data: 0,
				};
			}
			return {
				error: new UnexpectedMailchainError('Failed to get latest nonce.', error as Error),
			};
		}
	}
}
