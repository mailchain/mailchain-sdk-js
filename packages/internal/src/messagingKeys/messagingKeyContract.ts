import {
	ContractCall,
	createAxiosConfiguration,
	MessagingKeysApiFactory,
	MessagingKeysApiInterface,
} from '@mailchain/api';
import axios, { AxiosInstance } from 'axios';
import { ETHEREUM, NEAR, TEZOS, ProtocolType } from '@mailchain/addressing';
import { convertPublic } from '@mailchain/api/helpers/apiKeyToCryptoKey';
import { MessagingKeyVerificationError } from '@mailchain/signatures';
import { PublicKey } from '@mailchain/crypto';
import { MAILCHAIN, ProtocolNotSupportedError } from '@mailchain/addressing/protocols';
import { Configuration, MailchainResult } from '../';
import { UnexpectedMailchainError } from '../errors';
import { NearContractCallResolver } from './contractResolvers/near';
import { ContractCallMessagingKeyResolver } from './contractResolvers/resolver';
import { MessagingKeyVerifier } from './verify';
import { MailchainKeyRegContractCallResolver } from './contractResolvers/mailchain';
import { InvalidContractResponseError, MessagingKeyNotFoundInContractError } from './contractResolvers/errors';
import { ResolveAddressError, RegisteredResolvedAddress, VendedResolvedAddress } from './messagingKeys';
import { MessagingKeyContactError } from './errors';

export type ContractCallResolvedAddress =
	| Omit<RegisteredResolvedAddress, 'mailchainAddress'>
	| Omit<VendedResolvedAddress, 'mailchainAddress'>;

export class MessagingKeyContractCall {
	constructor(
		private readonly resolvers: Map<ProtocolType, ContractCallMessagingKeyResolver>,
		private readonly messagingKeysApi: MessagingKeysApiInterface,
		private readonly messagingKeyVerifier: MessagingKeyVerifier,
	) {}

	static create(configuration: Configuration, axiosInstance: AxiosInstance = axios.create()) {
		const mailchainKeyRegistryResolver = MailchainKeyRegContractCallResolver.create(configuration, axiosInstance);
		return new MessagingKeyContractCall(
			new Map<ProtocolType, ContractCallMessagingKeyResolver>([
				[ETHEREUM, mailchainKeyRegistryResolver],
				[MAILCHAIN, mailchainKeyRegistryResolver],
				[NEAR, NearContractCallResolver.create(configuration, axiosInstance)],
				[TEZOS, mailchainKeyRegistryResolver],
			]),
			MessagingKeysApiFactory(createAxiosConfiguration(configuration.apiPath)),
			MessagingKeyVerifier.create(configuration),
		);
	}

	async resolve(
		contractCall: ContractCall,
		identityKey?: PublicKey,
	): Promise<MailchainResult<ContractCallResolvedAddress, ResolveAddressError>> {
		const protocol = contractCall.protocol as ProtocolType;
		const resolver = this.resolvers.get(protocol);
		if (!resolver) {
			return { error: new ProtocolNotSupportedError(protocol) };
		}

		const { data, error } = await resolver.resolve(contractCall);
		if (error instanceof MessagingKeyNotFoundInContractError) {
			return this.getVerifiedVendedPublicMessagingKey(contractCall.address, protocol, identityKey);
		} else if (error instanceof InvalidContractResponseError) {
			return { error: new MessagingKeyContactError(error) };
		} else if (error) {
			return {
				error: new MessagingKeyContactError(error),
			};
		}

		if (!identityKey) {
			return {
				error: new UnexpectedMailchainError('identityKey is required'),
			};
		}

		const { messagingKey, proof } = data;
		return {
			data: {
				type: 'registered',
				messagingKey,
				identityKey,
				protocol,
				protocolAddress: contractCall.address,
				proof,
			},
		};
	}

	private async getVerifiedVendedPublicMessagingKey(
		address: string,
		protocol: ProtocolType,
		identityKey?: PublicKey,
	): Promise<MailchainResult<ContractCallResolvedAddress, MessagingKeyVerificationError>> {
		const vendedKeyResponse = await this.messagingKeysApi.getVendedPublicMessagingKey(address, protocol as any);
		const verified = await this.messagingKeyVerifier.verifyProvidedKeyProof(
			vendedKeyResponse.data.proof,
			convertPublic(vendedKeyResponse.data.messagingKey),
		);
		if (!verified) {
			return { error: new MessagingKeyVerificationError() };
		}
		return {
			data: {
				messagingKey: convertPublic(vendedKeyResponse.data.messagingKey),
				identityKey,
				protocol,
				type: 'vended',
				protocolAddress: address,
			},
		};
	}
}
