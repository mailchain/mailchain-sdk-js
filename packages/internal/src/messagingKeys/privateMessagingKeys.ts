import { PrivateKey, isPublicKeyEqual } from '@mailchain/crypto';
import { KeyRing } from '@mailchain/keyring';
import { ProvidedMessagingKeyIncorrectError } from '@mailchain/signatures';
import { MailchainResult } from '..';
import { Configuration } from '../configuration';
import { AddressNonce, GetMessagingKeyLatestNonceError } from './addressNonce';
import { MessagingKeys, ResolveAddressError } from './messagingKeys';
import { AddressNotRegisteredError } from './errors';

export type GetExportablePrivateMessagingKeyError =
	| ResolveAddressError
	| GetMessagingKeyLatestNonceError
	| ProvidedMessagingKeyIncorrectError
	| AddressNotRegisteredError;

export class PrivateMessagingKeys {
	constructor(private readonly messagingKeys: MessagingKeys, private readonly addressNonce: AddressNonce) {}

	static create(config: Configuration) {
		return new PrivateMessagingKeys(MessagingKeys.create(config), AddressNonce.create(config));
	}

	async getExportablePrivateMessagingKey(
		address: string,
		keyRing: KeyRing,
	): Promise<MailchainResult<PrivateKey, GetExportablePrivateMessagingKeyError>> {
		const { data: resolvedAddress, error: resolveAddressError } = await this.messagingKeys.resolve(address);

		if (resolveAddressError != null) {
			return { error: resolveAddressError };
		}

		const { data: getMessagingKeyLatestNonce, error: getMessagingKeyLatestNonceError } =
			await this.addressNonce.getMessagingKeyLatestNonce(
				resolvedAddress.protocolAddress,
				resolvedAddress.protocol,
			);

		if (getMessagingKeyLatestNonceError != null) {
			return { error: getMessagingKeyLatestNonceError };
		}

		if (resolvedAddress.type === 'vended') {
			return { error: new AddressNotRegisteredError() };
		}

		const privateMessagingKey = keyRing.addressExportableMessagingKey(
			resolvedAddress.protocolAddress,
			resolvedAddress.protocol,
			getMessagingKeyLatestNonce,
		);

		if (!isPublicKeyEqual(privateMessagingKey.publicKey, resolvedAddress.messagingKey)) {
			return { error: new ProvidedMessagingKeyIncorrectError('address') };
		}

		return { data: privateMessagingKey };
	}
}
