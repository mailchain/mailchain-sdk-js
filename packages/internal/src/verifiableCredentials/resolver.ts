import { encodeHex } from '@mailchain/encoding';
import { DIDResolutionOptions, DIDResolutionResult, Resolvable } from 'did-resolver';
import { MessagingKeys, ResolvedAddress } from '../messagingKeys';
import { Configuration } from '../configuration';

export class MailchainDIDMessagingKeyResolver implements Resolvable {
	constructor(private readonly messagingKeys: MessagingKeys) {}
	static create(configuration: Configuration) {
		return new MailchainDIDMessagingKeyResolver(MessagingKeys.create(configuration));
	}
	async resolve(didUrl: string, _options?: DIDResolutionOptions): Promise<DIDResolutionResult> {
		const address = didUrl.replace('did:mailchain:', '');
		const { data: resolvedAddress, error: resolveAddressError } = await this.messagingKeys.resolve(address);
		if (resolveAddressError) {
			throw resolveAddressError;
		}

		return createDidDocumentFromResolvedAddress(didUrl, resolvedAddress);
	}
}

export function createDidDocumentFromResolvedAddress(
	didUrl: string,
	resolvedAddress: ResolvedAddress,
): DIDResolutionResult {
	return {
		didDocument: {
			id: didUrl,
			authentication: [
				{
					id: `${didUrl}/messaging-key`,
					controller: didUrl,
					type: 'ED25519SignatureVerification',
					publicKeyHex: `${encodeHex(resolvedAddress.messagingKey.bytes)}`,
				},
			],
			verificationMethod: [], // no verification methods
		},
		didDocumentMetadata: {
			canonicalId: didUrl,
		},
		didResolutionMetadata: {
			contentType: 'application/did+ld+json',
		},
	};
}
