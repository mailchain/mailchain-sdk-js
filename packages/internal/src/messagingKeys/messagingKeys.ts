import { encodeHexZeroX } from '@mailchain/encoding';
import { ALL_PROTOCOLS, encodeAddressByProtocol, ProtocolNotSupportedError, ProtocolType } from '@mailchain/addressing';
import { encodePublicKey, PublicKey } from '@mailchain/crypto';
import {
	AddressesApiFactory,
	createAxiosConfiguration,
	IdentityKeysApiFactory,
	CryptoKeyConvert,
	encodingTypeToEncodingEnum,
	IdentityKeysApiInterface,
	AddressesApiInterface,
	GetAddressMessagingKeyResponseBody,
} from '@mailchain/api';
import { convertPublic } from '@mailchain/api/helpers/apiKeyToCryptoKey';
import { MessagingKeyVerificationError } from '@mailchain/signatures';
import { isAxiosError } from 'axios';
import { Configuration } from '../configuration';
import { MailchainResult, partitionMailchainResults } from '../';
import {
	MessagingKeyContactError,
	NameserviceAddressNotFoundError,
	NameserviceAddressUnresolvableError,
	UnexpectedMailchainError,
} from './errors';
import { MessagingKeyProof } from './proof';
import { MessagingKeyContractCall } from './messagingKeyContract';

type BaseResolvedAddress = {
	/** Messaging key to be used when communicate with the resolved address. See {@link PublicKey} */
	messagingKey: PublicKey;
	/** Protocol of resolved address. Protocol is determined when resolving the address. */
	protocol: ProtocolType;
	/** Identity key of resolved address. Identity key might be undefined in the case of `vended` messaging key. */
	identityKey?: PublicKey;
	/** Protocol address that has been resolved in the case of name services. */
	protocolAddress: string;
};

export type RegisteredResolvedAddress = BaseResolvedAddress & {
	/** Indicates the messaging key has been registered by a user. */
	type: 'registered';
};

export type VendedResolvedAddress = BaseResolvedAddress & {
	/** Indicates the messaging key has been vended by Mailchain. */
	type: 'vended';
};

/**
 * Resolved address response containing a proven messaging key.
 */
export type ResolvedAddress = RegisteredResolvedAddress | VendedResolvedAddress;
export type ResolveAddressError =
	| ProtocolNotSupportedError
	| MessagingKeyContactError
	| MessagingKeyVerificationError
	| NameserviceAddressNotFoundError
	| NameserviceAddressUnresolvableError
	| UnexpectedMailchainError;
export type ResolveAddressResult = MailchainResult<ResolvedAddress, ResolveAddressError>;

export type ResolvedManyAddresses = Map<string, ResolvedAddress>;
export type ResolvedManyAddressesResult = MailchainResult<ResolvedManyAddresses>;
export type ResolveManyAddressesError = SomeAddressesUnresolvableError;

export class SomeAddressesUnresolvableError extends Error {
	readonly type = 'not_all_addresses_resolved';
	constructor(
		public readonly successes: Array<{ params: string; data: ResolvedAddress }>,
		public readonly failures: Array<{ params: string; error: ResolveAddressError }>,
	) {
		super(`Not all addresses were resolved. Check the failed resolutions for more information.`);
	}
}

export class MessagingKeys {
	constructor(
		private readonly addressApi: AddressesApiInterface,
		private readonly identityKeysApi: IdentityKeysApiInterface,
		private readonly messagingKeyContractCall: MessagingKeyContractCall,
	) {}

	static create(configuration: Configuration) {
		return new MessagingKeys(
			AddressesApiFactory(createAxiosConfiguration(configuration.apiPath)),
			IdentityKeysApiFactory(createAxiosConfiguration(configuration.apiPath)),
			MessagingKeyContractCall.create(configuration),
		);
	}

	/**
	 * Resolve the messaging key for the given address.
	 *
	 * @param address Address to resolve.
	 *
	 * @returns A {@link ResolvedAddress resolved address} which may be a registered or vended.
	 *
	 * @example
	 * import { messagingKeys } from '@mailchain/sdk';
	 *
	 * const resolvedAddress = await messagingKeys.resolve(address);
	 * console.log(resolvedAddress);
	 *
	 */
	async resolve(address: string): Promise<ResolveAddressResult> {
		const { data, error } = await this.getAddressMessagingKey(address);
		if (error != null) {
			return { error };
		}

		return this.messagingKeyContractCall.resolve(
			data.protocol as ProtocolType,
			data.contractCall,
			data.identityKey ? convertPublic(data.identityKey) : undefined,
		);
	}

	async resolveMany(addresses: string[]): Promise<MailchainResult<ResolvedManyAddresses, ResolveManyAddressesError>> {
		const deduplicatedAddresses = [...new Set(addresses)];
		const resolvedAddresses = await Promise.all(
			deduplicatedAddresses.map(async (address) => {
				const resolvedAddress = await this.resolve(address);
				return { params: address, result: resolvedAddress };
			}),
		);

		const { failures, successes } = partitionMailchainResults(resolvedAddresses);

		if (failures.length > 0) {
			return { error: new SomeAddressesUnresolvableError(successes, failures) };
		}

		return {
			data: new Map(successes.map((r) => [r.params, r.data])),
		};
	}

	async update(proof: MessagingKeyProof): Promise<void> {
		const encodedIdentityKey = encodeHexZeroX(encodePublicKey(proof.identityKey));
		const encodedAddress = encodeAddressByProtocol(proof.address, proof.protocol);

		await this.identityKeysApi.putMsgKeyByIDKey(encodedIdentityKey, {
			address: {
				encoding: encodingTypeToEncodingEnum(encodedAddress.encoding),
				value: encodedAddress.encoded,
				network: proof.network,
				protocol: proof.protocol,
			},
			locale: proof.locale,
			messageVariant: proof.messageVariant,
			messagingKey: CryptoKeyConvert.public(proof.messagingKey),
			nonce: proof.nonce,
			signature: encodeHexZeroX(proof.signature),
			signatureMethod: proof.signatureMethod,
		});
	}

	private async getAddressMessagingKey(
		address: string,
	): Promise<
		MailchainResult<
			GetAddressMessagingKeyResponseBody,
			| ProtocolNotSupportedError
			| UnexpectedMailchainError
			| NameserviceAddressNotFoundError
			| NameserviceAddressUnresolvableError
		>
	> {
		try {
			const { data } = await this.addressApi.getAddressMessagingKey(address);

			const protocol = data.protocol as ProtocolType;
			if (!ALL_PROTOCOLS.includes(protocol)) {
				return { error: new ProtocolNotSupportedError(protocol) };
			}

			return { data };
		} catch (e) {
			if (isAxiosError(e)) {
				switch (e.response?.status) {
					case 404:
						return {
							error: new NameserviceAddressNotFoundError(),
						};
					case 422:
						return {
							error: new NameserviceAddressUnresolvableError(),
						};
				}
			}
			return {
				error: new UnexpectedMailchainError(`Failed to resolve messaging key of address ${address}`, e),
			};
		}
	}
}
