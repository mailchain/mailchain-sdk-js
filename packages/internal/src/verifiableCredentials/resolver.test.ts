import { AliceSECP256K1PublicAddressStr } from '@mailchain/addressing/protocols/ethereum/test.const';
import { mock } from 'jest-mock-extended';
import { AliceED25519PublicKey } from '@mailchain/crypto/ed25519/test.const';
import { AliceSECP256K1PublicKey } from '@mailchain/crypto/secp256k1/test.const';
import { MessagingKeys, Proof } from '../messagingKeys';
import { MailchainDIDMessagingKeyResolver } from './resolver';

describe('MailchainDIDMessagingKeyResolver', () => {
	it('should resolve', async () => {
		const mockMessagingKeys = mock<MessagingKeys>();
		mockMessagingKeys.resolveIndividual.mockResolvedValue({
			data: {
				mailchainAddress: `${AliceSECP256K1PublicAddressStr}@ethereum.mailchain.com`,
				messagingKey: AliceED25519PublicKey,
				protocol: 'ethereum',
				protocolAddress: AliceSECP256K1PublicAddressStr,
				type: 'registered',
				identityKey: AliceSECP256K1PublicKey,
				proof: {} as Proof,
			},
		});

		const target = new MailchainDIDMessagingKeyResolver(mockMessagingKeys);
		const result = await target.resolve(`did:mailchain:${AliceSECP256K1PublicAddressStr}@ethereum.mailchain.com`);
		expect(result).toMatchSnapshot('result');
		expect(mockMessagingKeys.resolveIndividual).toHaveBeenCalledWith(
			`${AliceSECP256K1PublicAddressStr}@ethereum.mailchain.com`,
		);
	});
});
