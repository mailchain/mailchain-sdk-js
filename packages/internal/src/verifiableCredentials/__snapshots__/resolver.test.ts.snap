// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MailchainDIDMessagingKeyResolver should resolve: result 1`] = `
{
  "didDocument": {
    "authentication": [
      {
        "controller": "did:mailchain:0xD5ab4CE3605Cd590Db609b6b5C8901fdB2ef7FE6@ethereum.mailchain.com",
        "id": "did:mailchain:0xD5ab4CE3605Cd590Db609b6b5C8901fdB2ef7FE6@ethereum.mailchain.com/messaging-key",
        "publicKeyHex": "723caa23a5b511af5ad7b7ef6076e414ab7e75a9dc910ea60e417a2b770a5671",
        "type": "ED25519SignatureVerification",
      },
    ],
    "id": "did:mailchain:0xD5ab4CE3605Cd590Db609b6b5C8901fdB2ef7FE6@ethereum.mailchain.com",
    "verificationMethod": [],
  },
  "didDocumentMetadata": {
    "canonicalId": "did:mailchain:0xD5ab4CE3605Cd590Db609b6b5C8901fdB2ef7FE6@ethereum.mailchain.com",
  },
  "didResolutionMetadata": {
    "contentType": "application/did+ld+json",
  },
}
`;
